### PR TITLE
Playlist hls support

### PR DESCRIPTION
**Pull request by deeppandya**
Resolves https://github.com/brave/brave-browser/issues/31001
Resolves https://github.com/brave/brave-browser/issues/30682
Resolves  https://github.com/brave/brave-browser/issues/31716
Resolves https://github.com/brave/brave-browser/issues/29418
Resolves https://github.com/brave/brave-browser/issues/30569
Resolves https://github.com/brave/brave-browser/issues/30567
Resolves https://github.com/brave/brave-browser/issues/30568
Resolves https://github.com/brave/brave-browser/issues/30682

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: android/java/org/chromium/chrome/browser/toolbar/top/BraveToolbarLayoutImpl.java [hunk 1]
```diff
@@ -1739,9 +1717,20 @@ public void drawAnimationOverlay(ViewGroup toolbarButtonsContainer, Canvas canva
     }
 
     @Override
-    public void onEvent(int eventType, String playlistId) {
+    public void onEvent(int eventType, String id) {
         if (eventType == PlaylistEvent.ITEM_ADDED) {
             showAddedToPlaylistSnackBar();
+            if (mPlaylistService != null) {
+            }
```
Comment (SergeyZhukovsky) [inline android/java/org/chromium/chrome/browser/toolbar/top/BraveToolbarLayoutImpl.java:1724]: should we have a return if the service is null?

File: third_party/android_deps/BUILD.gn [hunk 2]
```diff
@@ -205,6 +207,16 @@ android_aar_prebuilt(
   ]
 }
 
+android_aar_prebuilt("com_google_android_exoplayer_exoplayer_hls_java") {
+  aar_path =
+      "libs/com_google_android_exoplayer_exoplayer_hls/exoplayer-hls-2.17.1.aar"
```
[the remark this answers is not in the document]
Comment (deeppandya) [reply third_party/android_deps/BUILD.gn:212]: i had this review for all exoplayer deps here : https://github.com/brave/security/issues/1068

[the remark this answers is not in the document]
Comment (deeppandya) [reply third_party/android_deps/BUILD.gn:212]: also created new review for this specific deps : https://github.com/brave/security/issues/1308

**Review by SergeyZhukovsky — APPROVED** [review]
++

**Review by mkarolin — APPROVED** [review]
`strings`++

**Review by sujitacharya2005 — APPROVED** [review]
LGTM

File: android/java/AndroidManifest.xml [hunk 3]
```diff
@@ -217,9 +217,16 @@
     android:exported="false"
     android:launchMode="singleTask"
     android:theme="@style/Theme.Chromium.Activity.Fullscreen" />
+
 <service android:name="com.brave.playlist.PlaylistVideoService"
     android:exported="false"/>
 
+    <service
+            android:name="com.brave.playlist.PlaylistDownloadService"
```
Comment (stoletheminerals) [inline android/java/AndroidManifest.xml:225]: This Service is using exoplayer lib to download files **https://github.com/brave-experiments/playlist-kotlin/blob/da5a721c0ea0c99ca2d2ad0015af988a8afce647/src/main/java/com/brave/playlist/PlaylistDownloadUtils.kt#L20.** 

We should use Chromium's network stack instead:
https://github.com/brave/security/issues/1202#issuecomment-1450252466

**Review by stoletheminerals — CHANGES_REQUESTED** [review]

**Review by tapanmodh — APPROVED** [review]
lgtm

File: third_party/android_deps/libs/com_google_android_exoplayer_exoplayer_hls/README.chromium [hunk 4]
```diff
@@ -0,0 +1,4 @@
+Name: Exoplayer HLS module
+URL: https://github.com/google/ExoPlayer/tree/release-v2/library/hls
+License: Apache-2.0
+License File: /brave/common/licenses/Apache-2.0
```
Comment (tapanmodh) [inline third_party/android_deps/libs/com_google_android_exoplayer_exoplayer_hls/README.chromium:4]: new line please

**Comment by kjozwiak** [comment]
Removed the `1.57.x` milestone as the above wasn't merged into `master` yet which is currently on `1.58.x`. Once this gets merged into `master`, it will be moved into the appropriate milestone.

**Comment by bsclifton** [comment]
Closing - this feature was PRed with https://github.com/brave/brave-core/pull/21041 and merged 😄👍

**Pull request closed without merging**